File: src/main/java/io/github/oleksivio/telegram/bot/api/annotations/filter/telegram/AudioFilter.java
package io.github.oleksivio.telegram.bot.api.annotations.filter.telegram;

import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.IntegerFilter;
import io.github.oleksivio.telegram.bot.api.annotations.filter.primitive.StringFilter;
import io.github.oleksivio.telegram.bot.api.model.annotation.AnnotationState;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface AudioFilter {

    AnnotationState status() default AnnotationState.ON;

    String[] validator() default {};

    IntegerFilter duration() default @IntegerFilter(status = AnnotationState.OFF);

    StringFilter performer() default @StringFilter(status = AnnotationState.OFF);

    StringFilter title() default @StringFilter(status = AnnotationState.OFF);

    StringFilter mimeType() default @StringFilter(status = AnnotationState.OFF);

    IntegerFilter fileSize() default @IntegerFilter(status = AnnotationState.OFF);

    PhotoFilter thumb() default @PhotoFilter(status = AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/AudioFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.IntegerFilter
import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class AudioFilter(
    val status: AnnotationState = AnnotationState.ON,
    val validator: Array<String> = [],
    val duration: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    ),
    val performer: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val title: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val mimeType: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val fileSize: IntegerFilter = IntegerFilter(
        status = AnnotationState.OFF
    ),
    val thumb: PhotoFilter = PhotoFilter(
        status = AnnotationState.OFF
    )
)
